localmetrics/exporter.go

```diff
@@ -0,0 +1,164 @@
+package localmetrics
+
+import (
+	"context"
+	"fmt"
+
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/prometheus/common/log"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	portMetrics    = "metrics"
+	metricsPath    = "/metrics"
+	metricsPort    = "8383"
+	scrapeInterval = "1m"
+)
+
+type Exporter struct {
+	Client    client.Client
+	Name      string
+	Namespace string
+	OwnerRefs []metav1.OwnerReference
+	Labels    map[string]string
+}
+
+func NewExporter(client client.Client, name, namespace string, ownerRefs []metav1.OwnerReference, labels map[string]string) *Exporter {
+	return &Exporter{
+		Client:    client,
+		Name:      name,
+		Namespace: namespace,
+		OwnerRefs: ownerRefs,
+		Labels:    labels,
+	}
+}
+
+// EnableMetricsExporter is a wrapper around createOrUpdateService()
+// and createOrUpdateServiceMonitor()
+func (e *Exporter) EnableMetricsExporter() error {
+	_, err := e.createOrUpdateService()
+	if err != nil {
+		return err
+	}
+	_, err = e.createOrUpdateServiceMonitor()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e *Exporter) getMetricsExporterService() *corev1.Service {
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            e.Name,
+			Namespace:       e.Namespace,
+			Labels:          e.Labels,
+			OwnerReferences: e.OwnerRefs,
+		},
+		Spec: corev1.ServiceSpec{
+			Type: corev1.ServiceTypeClusterIP,
+			Ports: []corev1.ServicePort{
+				{
+					Name:     portMetrics,
+					Port:     int32(8383),
+					Protocol: corev1.ProtocolTCP,
+					TargetPort: intstr.IntOrString{
+						Type:   intstr.Int,
+						IntVal: int32(8383),
+						StrVal: "8383",
+					},
+				},
+			},
+			Selector: e.Labels,
+		},
+	}
+	return service
+}
+
+// createOrUpdateService creates service object or an error
+func (e *Exporter) createOrUpdateService() (*corev1.Service, error) {
+	service := e.getMetricsExporterService()
+	namespacedName := types.NamespacedName{Namespace: service.GetNamespace(), Name: service.GetName()}
+
+	log.Info("Reconciling metrics exporter service", "NamespacedName", namespacedName)
+
+	oldService := &corev1.Service{}
+	err := e.Client.Get(context.TODO(), namespacedName, oldService)
```

Comment:
I think we should not be using `context.TODO` everywhere. Can we refactor this code to use proper contexts?

Reply:
using proper context from the daemonset reconciler.